// @imtbl/zkevm-contracts/dist/typechain-types/factories/@openzeppelin/contracts/utils/ShortStrings__factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShortStrings__factory = void 0;
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [],
        name: "InvalidShortString",
        type: "error",
    },
    {
        inputs: [
            {
                internalType: "string",
                name: "str",
                type: "string",
            },
        ],
        name: "StringTooLong",
        type: "error",
    },
];
const _bytecode = "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea2646970667358221220d46f6bc23c7b4f554fa107f601abd879947bf1b97495af2919c114ef5ad3e5c064736f6c63430008110033";
const isSuperArgs = (xs) => xs.length > 1;
class ShortStrings__factory extends ethers_1.ContractFactory {
    constructor(...args) {
        if (isSuperArgs(args)) {
            super(...args);
        }
        else {
            super(_abi, _bytecode, args[0]);
        }
    }
    deploy(overrides) {
        return super.deploy(overrides || {});
    }
    getDeployTransaction(overrides) {
        return super.getDeployTransaction(overrides || {});
    }
    attach(address) {
        return super.attach(address);
    }
    connect(signer) {
        return super.connect(signer);
    }
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.ShortStrings__factory = ShortStrings__factory;
ShortStrings__factory.bytecode = _bytecode;
ShortStrings__factory.abi = _abi;
