// @imtbl/zkevm-contracts/dist/typechain-types/factories/contracts/errors/Errors.sol/OperatorAllowlistEnforcementErrors__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OperatorAllowlistEnforcementErrors__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [],
        name: "AllowlistDoesNotImplementIOperatorAllowlist",
        type: "error",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "target",
                type: "address",
            },
        ],
        name: "ApproveTargetNotInAllowlist",
        type: "error",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "approver",
                type: "address",
            },
        ],
        name: "ApproverNotInAllowlist",
        type: "error",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "caller",
                type: "address",
            },
        ],
        name: "CallerNotInAllowlist",
        type: "error",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "from",
                type: "address",
            },
        ],
        name: "TransferFromNotInAllowlist",
        type: "error",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "to",
                type: "address",
            },
        ],
        name: "TransferToNotInAllowlist",
        type: "error",
    },
];
class OperatorAllowlistEnforcementErrors__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.OperatorAllowlistEnforcementErrors__factory = OperatorAllowlistEnforcementErrors__factory;
OperatorAllowlistEnforcementErrors__factory.abi = _abi;
