// @opensea/seaport-js/lib/utils/eip712/defaults.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DefaultGetter = void 0;
const logger_1 = require("@ethersproject/logger");
const utils_1 = require("ethers/lib/utils");
const logger = new logger_1.Logger("defaults");
const baseDefaults = {
    integer: 0,
    address: (0, utils_1.hexZeroPad)("0x", 20),
    bool: false,
    bytes: "0x",
    string: "",
};
const isNullish = (value) => {
    if (value === undefined)
        return false;
    return (value !== undefined &&
        value !== null &&
        ((["string", "number"].includes(typeof value) &&
            BigInt(value) === BigInt(0)) ||
            (Array.isArray(value) && value.every(isNullish)) ||
            (typeof value === "object" && Object.values(value).every(isNullish)) ||
            (typeof value === "boolean" && value === false)));
};
function getDefaultForBaseType(type) {
    var _a;
    // bytesXX
    const [, width] = (_a = type.match(/^bytes(\d+)$/)) !== null && _a !== void 0 ? _a : [];
    if (width)
        return (0, utils_1.hexZeroPad)("0x", parseInt(width));
    if (type.match(/^(u?)int(\d*)$/))
        type = "integer";
    return baseDefaults[type];
}
class DefaultGetter {
    constructor(types) {
        this.types = types;
        this.defaultValues = {};
        for (const name in types) {
            const defaultValue = this.getDefaultValue(name);
            this.defaultValues[name] = defaultValue;
            if (!isNullish(defaultValue)) {
                logger.throwError(`Got non-empty value for type ${name} in default generator: ${defaultValue}`);
            }
        }
    }
    static from(types, type) {
        const { defaultValues } = new DefaultGetter(types);
        if (type)
            return defaultValues[type];
        return defaultValues;
    }
    /* eslint-enable no-dupe-class-members */
    getDefaultValue(type) {
        if (this.defaultValues[type])
            return this.defaultValues[type];
        // Basic type (address, bool, uint256, etc)
        const basic = getDefaultForBaseType(type);
        if (basic !== undefined)
            return basic;
        // Array
        const match = type.match(/^(.*)(\x5b(\d*)\x5d)$/);
        if (match) {
            const subtype = match[1];
            const length = parseInt(match[3]);
            if (length > 0) {
                const baseValue = this.getDefaultValue(subtype);
                return Array(length).fill(baseValue);
            }
            return [];
        }
        // Struct
        const fields = this.types[type];
        if (fields) {
            return fields.reduce((obj, { name, type }) => (Object.assign(Object.assign({}, obj), { [name]: this.getDefaultValue(type) })), {});
        }
        return logger.throwArgumentError(`unknown type: ${type}`, "type", type);
    }
}
exports.DefaultGetter = DefaultGetter;
