// @opensea/seaport-js/src/typechain-types/factories/seaport-core/src/lib/AmountDeriver__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  AmountDeriver,
  AmountDeriverInterface,
} from "../../../../seaport-core/src/lib/AmountDeriver";

const _abi = [
  {
    inputs: [],
    name: "InexactFraction",
    type: "error",
  },
] as const;

const _bytecode =
  "0x6080604052348015600f57600080fd5b50603f80601d6000396000f3fe6080604052600080fdfea2646970667358221220ff1f7fc7f59825e8cfae880e6b76756a33b588f7daa78087627017de75dc580064736f6c63430008110033";

type AmountDeriverConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: AmountDeriverConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class AmountDeriver__factory extends ContractFactory {
  constructor(...args: AmountDeriverConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<AmountDeriver> {
    return super.deploy(overrides || {}) as Promise<AmountDeriver>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): AmountDeriver {
    return super.attach(address) as AmountDeriver;
  }
  override connect(signer: Signer): AmountDeriver__factory {
    return super.connect(signer) as AmountDeriver__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): AmountDeriverInterface {
    return new utils.Interface(_abi) as AmountDeriverInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): AmountDeriver {
    return new Contract(address, _abi, signerOrProvider) as AmountDeriver;
  }
}
