// @opensea/seaport-js/src/typechain-types/factories/seaport-core/src/lib/ZoneInteraction__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  ZoneInteraction,
  ZoneInteractionInterface,
} from "../../../../seaport-core/src/lib/ZoneInteraction";

const _abi = [
  {
    inputs: [
      {
        internalType: "bytes32",
        name: "orderHash",
        type: "bytes32",
      },
    ],
    name: "InvalidContractOrder",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "bytes32",
        name: "orderHash",
        type: "bytes32",
      },
    ],
    name: "InvalidRestrictedOrder",
    type: "error",
  },
] as const;

const _bytecode =
  "0x6080604052348015600f57600080fd5b50603f80601d6000396000f3fe6080604052600080fdfea264697066735822122088c7d715d9ce154b1396936dd1730f836078f506be7ba4b0a145a84488d7e26564736f6c63430008110033";

type ZoneInteractionConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: ZoneInteractionConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class ZoneInteraction__factory extends ContractFactory {
  constructor(...args: ZoneInteractionConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<ZoneInteraction> {
    return super.deploy(overrides || {}) as Promise<ZoneInteraction>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): ZoneInteraction {
    return super.attach(address) as ZoneInteraction;
  }
  override connect(signer: Signer): ZoneInteraction__factory {
    return super.connect(signer) as ZoneInteraction__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): ZoneInteractionInterface {
    return new utils.Interface(_abi) as ZoneInteractionInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): ZoneInteraction {
    return new Contract(address, _abi, signerOrProvider) as ZoneInteraction;
  }
}
